lib/Dialect/ESI/ESIOps.cpp

```diff
@@ -126,6 +127,48 @@ void WrapValidReady::build(OpBuilder &b, OperationState &state, Value data,
         b.getI1Type(), data, valid);
 }
 
+namespace {
+/// Eliminate back-to-back wrap-unwraps to reduce the number of ESI channels.
+struct RemoveWrapUnwrap : public OpConversionPattern<WrapValidReady> {
+public:
+  RemoveWrapUnwrap(MLIRContext *ctxt) : OpConversionPattern(ctxt) {}
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult match(Operation *op) const final {
+    auto wrap = dyn_cast<WrapValidReady>(op);
+    if (!wrap)
+      return failure();
+    if (!wrap.chanOutput().hasOneUse())
+      return failure();
+    if (!isa<UnwrapValidReady>(wrap.chanOutput().use_begin()->getOwner()))
+      return failure();
+    return success();
+  }
+
+  void rewrite(WrapValidReady wrap, ArrayRef<Value> operands,
+               ConversionPatternRewriter &rewriter) const final;
+};
+} // anonymous namespace
+
+void RemoveWrapUnwrap::rewrite(WrapValidReady wrap,
+                               ArrayRef<Value> stageOperands,
+                               ConversionPatternRewriter &rewriter) const {
+  // This pass requires that implicit forks already be lowered by the
+  // abstract-to-physical pass. Meaning exactly one user.
+  UnwrapValidReady unwrap =
+      dyn_cast<UnwrapValidReady>(wrap.chanOutput().use_begin()->getOwner());
+
+  rewriter.replaceOp(unwrap, {wrap.rawInput(), wrap.valid()});
+  rewriter.replaceOp(wrap, {unwrap.ready(), wrap.ready()});
+  rewriter.eraseOp(unwrap);
```

Comment:
@teqdruid shot in the dark, but is it a problem to erase ops that have already been replaced? I think you don't need to call eraseOp, but not sure if that's the issue

Reply:
Good catch, but that wasn't it.

Reply:
Oh and upgrading LLVM didn't help.